libclamav/crtmgr.c

```diff
@@ -281,12 +325,133 @@ void crtmgr_free(crtmgr *m)
         crtmgr_del(m, m->crts);
 }
 
-static int crtmgr_rsa_verify(cli_crt *x509, fp_int *sig, cli_crt_hashtype hashtype, const uint8_t *refhash)
+static cl_error_t _padding_check_PKCS1_type_1(uint8_t **to, int *tlen,
+                                              uint8_t *from, unsigned int flen,
+                                              unsigned int num)
+{
+    int i, j;
+    unsigned char *p;
+
+    p = from;
+
+    /*
+     * The format is
+     * 00 || 01 || PS || 00 || D
+     * PS - padding string, at least 8 bytes of FF
+     * D  - data.
+     */
+
+    if (num < 11) /* RSA_PKCS1_PADDING_SIZE */
+        return CL_EPARSE;
+
+    /* Accept inputs with and without the leading 0-byte. */
+    if (num == flen) {
+        if ((*p++) != 0x00) {
+            cli_dbgmsg("%s: Bad padding\n", __func__);
+            return CL_EPARSE;
+        }
+        flen--;
+    }
+
+    if ((num != (flen + 1)) || (*(p++) != 0x01)) {
+        cli_dbgmsg("%s: Bad block type\n", __func__);
+        return CL_EPARSE;
+    }
+
+    /* scan over padding data */
+    j = flen - 1; /* one for type. */
+    for (i = 0; i < j; i++) {
+        if (*p != 0xff) { /* should decrypt to 0xff */
+            if (*p == 0) {
+                p++;
+                break;
+            } else {
+                cli_dbgmsg("%s: Bad header\n", __func__);
+                return CL_EPARSE;
+            }
+        }
+        p++;
+    }
+
+    if (i == j) {
+        cli_dbgmsg("%s: Bad header\n", __func__);
+        return CL_EPARSE;
+    }
+
+    if (i < 8) {
+        cli_dbgmsg("%s: Bad padding\n", __func__);
+        return CL_EPARSE;
+    }
+    i++; /* Skip over the '\0' */
+    j -= i;
+    *tlen = j;
+    *to   = p;
+
+    return CL_SUCCESS;
+}
+
+static cl_error_t crtmgr_get_recov_data(BIGNUM *sig, cli_crt *x509,
+                                        uint8_t **buffer, uint8_t **payload,
+                                        int *payload_len)
 {
-    int keylen = fp_unsigned_bin_size(&x509->n), siglen = fp_unsigned_bin_size(sig);
-    int ret, j, objlen, hashlen;
-    uint8_t d[513];
-    fp_int x;
+    BN_CTX *bnctx;
+    int pad_size;
+    int keylen;
+    uint8_t *d;
+    BIGNUM *x;
+    cl_error_t ret;
+
+    *buffer      = NULL;
+    *payload     = NULL;
+    *payload_len = 0;
+    ret          = CL_ERROR;
+
+    keylen = BN_num_bytes(x509->n);
+    bnctx  = BN_CTX_new();
+    if (!bnctx)
+        goto done;
+
+    x = BN_new();
+    if (!x)
+        goto done;
+
+    MALLOC(d, keylen);
+
+    if (!BN_mod_exp(x, sig, x509->e, x509->n, bnctx)) {
+        cli_warnmsg("crtmgr_rsa_verify: verification failed: BN_mod_exp failed.\n");
+        goto done;
+    }
+
+    pad_size = BN_bn2binpad(x, d, keylen);
```

Comment:
Can we accomplish the same thing using `BN_bn2bin` instead?  It seems `BN_bn2binpad` isn't present in openssl 1.0.2k, the version present on centos 7.